src/OpenSSL/crypto.py

```diff
@@ -217,11 +217,16 @@ def generate_key(self, type, bits):
                 _raise_current_error()
 
         elif type == TYPE_DSA:
-            dsa = _lib.DSA_generate_parameters(
-                bits, _ffi.NULL, 0, _ffi.NULL, _ffi.NULL, _ffi.NULL, _ffi.NULL)
+            dsa = _lib.DSA_new()
             if dsa == _ffi.NULL:
                 # TODO: This is untested.
                 _raise_current_error()
+            res = _lib.DSA_generate_parameters_ex(
+                dsa, bits, _ffi.NULL, 0, _ffi.NULL, _ffi.NULL, _ffi.NULL
+            )
+            if not res == 1:
+                # TODO: This is untested.
```

Comment:
I can register the DSA object to free with the gc (ffi.gc) but `assign_DSA` assigns memory ownership to the EVP_PKEY object, so that will ultimately free the memory in the success case. So either we explicitly free it in this path or I register it for gc at creation and then switch to `set1_DSA` which will increment refcount and make it so calling DSA_free the first time will decrement refcount by 1 (when the proxy object falls out of scope), and then the EVP_PKEY_free call will work later as it'll decrement refcount to 0.